src/doorman/TelnetProtocol.cpp

```diff
@@ -0,0 +1,172 @@
+
+static bool supports_ansi(const char *src) {
+    static const char *terms[] = {"xterm", "xterm-colour", "xterm-color", "ansi",      "xterm-ansi",
+                                  "vt100", "vt102",        "vt220",       "terminator"};
+    for (auto &term : terms)
+        if (!strcasecmp(src, term))
+            return true;
+    return false;
+}
+
+void TelnetProtocol::send_com(byte a, byte b) {
+    const byte buf[] = {IAC, a, b};
+    handler_.send_bytes(buf);
+}
+
+void TelnetProtocol::send_opt(byte a) {
+    const byte buf[] = {IAC, SB, a, TELQUAL_SEND, IAC, SE};
+    handler_.send_bytes(buf);
+}
+
+gsl::span<const byte> TelnetProtocol::interpret_iacs() {
+    auto starting_point = 0ul;
+    for (;;) {
+        auto iac_it = std::find(std::next(buffer_.begin(), starting_point), buffer_.end(), IAC);
+        if (iac_it == buffer_.end())
+            break;
+        // Next time we look for a command, start at this point. We're about to erase the IAC (if we parsed it).
+        starting_point = std::distance(buffer_.begin(), iac_it);
+        gsl::span<const byte> iac_command(&*iac_it, &*buffer_.end());
+        auto num_consumed = on_command(iac_command);
+        if (num_consumed == 0) {
+            // If there wasn't enough data for the IAC command, then stop right now and do no further IAC processing: we
+            // must await more data to interpret the command. However, process the rest of the buffer up to the IAC.
+            return gsl::span<const byte>(buffer_).first(starting_point);
+        }
+        // Remove the command.
+        buffer_.erase(std::next(buffer_.begin(), starting_point),
+                      std::next(buffer_.begin(), starting_point + num_consumed));
+    }
+    return buffer_;
+}
+
+void TelnetProtocol::add_data(gsl::span<const byte> data) {
+    buffer_.insert(buffer_.end(), data.begin(), data.end());
+
+    auto span_to_scan = interpret_iacs();
+    if (span_to_scan.empty())
+        return;
+
+    auto as_string_view = std::string_view(reinterpret_cast<const char *>(span_to_scan.data()), span_to_scan.size());
+    auto remaining_data = handle_whole_lines(as_string_view);
+
+    const auto num_consumed = as_string_view.size() - remaining_data.size();
+    buffer_.erase(buffer_.begin(), std::next(buffer_.begin(), num_consumed));
+}
+
+std::string_view TelnetProtocol::handle_whole_lines(std::string_view data) {
+    while (!data.empty()) {
+        // Handle a \n\r or \r\n sequence; even if it crossed a buffer boundary.
+        if ((previous_separator_ == '\n' && data.front() == '\r')
+            || (previous_separator_ == '\r' && data.front() == '\n')) {
+            data.remove_prefix(1);
+            previous_separator_ = 0;
+            continue;
+        }
+
+        auto cr_or_lf = data.find_first_of("\n\r");
+        if (cr_or_lf == std::string_view::npos)
+            break;
+
+        previous_separator_ = data[cr_or_lf];
+        handler_.on_line(data.substr(0, cr_or_lf));
```

Comment:
ok so substr() is  pos, pos+count, so the cr or lf char itself won't get included in the substring

Reply:
Right :). [substring](https://en.cppreference.com/w/cpp/string/basic_string_view/substr) is `[pos, pos + count)` so this doesn't include the `cr`. I think most (sane!) language substrings work this way (python, Java et al)